node/tests/exchange.rs

```diff
@@ -0,0 +1,503 @@
+
+    // request by height
+    let received_header_by_height = client.p2p().get_header_by_height(10).await.unwrap();
+    assert_eq!(server_headers[9], received_header_by_height);
+
+    // request by hash
+    let expected_header = &server_headers[15];
+    let received_header_by_hash = client
+        .p2p()
+        .get_header(expected_header.hash())
+        .await
+        .unwrap();
+    assert_eq!(expected_header, &received_header_by_hash);
+
+    // request genesis by height
+    let received_genesis = client.p2p().get_header_by_height(1).await.unwrap();
+    assert_eq!(server_headers.first().unwrap(), &received_genesis);
+
+    // request entire store range
+    let received_all_headers = client
+        .p2p()
+        .get_verified_headers_range(&received_genesis, 19)
+        .await
+        .unwrap();
+    assert_eq!(server_headers[1..], received_all_headers);
+
+    // reqest more headers than available in store
+    // TODO: this reflects _current_ behaviour. once sessions are implemented it'll keep retrying
+    // and this test will need to be changed
+    let out_of_bounds = client
+        .p2p()
+        .get_verified_headers_range(&received_genesis, 20)
+        .await
+        .unwrap_err();
+    assert!(matches!(
+        out_of_bounds,
+        P2pError::Exchange(ExchangeError::InvalidResponse)
+    ));
+
+    // request unknown hash
+    let unstored_header = header_generator.next_of(&server_headers[0]);
+    let unexpected_hash = client
+        .p2p()
+        .get_header(unstored_header.hash())
+        .await
+        .unwrap_err();
+    assert!(matches!(
+        unexpected_hash,
+        P2pError::Exchange(ExchangeError::HeaderNotFound)
+    ));
+
+    // request unknown height
+    let unexpected_height = client.p2p().get_header_by_height(21).await.unwrap_err();
+    assert!(matches!(
+        unexpected_height,
+        P2pError::Exchange(ExchangeError::HeaderNotFound)
+    ));
+}
+
+#[tokio::test]
+async fn client_server_invalid_requests() {
+    // Server Node
+    let server = Node::new(NodeConfig {
+        store: gen_filled_store(20).0,
+        ..listening_test_node_config()
+    })
+    .await
+    .unwrap();
+
+    // give server a sec to breathe, otherwise occiasionally client has problems with connecting
+    sleep(Duration::from_millis(100)).await;
+    let server_addrs = server.p2p().listeners().await.unwrap();
+
+    // Client node
+    let client = Node::new(NodeConfig {
+        p2p_bootstrap_peers: server_addrs.clone(),
+        ..test_node_config()
+    })
+    .await
+    .unwrap();
+
+    client.p2p().wait_connected().await.unwrap();
+
+    let none_data = client
+        .p2p()
+        .exchange_header_request(HeaderRequest {
+            data: None,
+            amount: 1,
+        })
+        .await
+        .unwrap_err();
+    assert!(matches!(
+        none_data,
+        P2pError::Exchange(ExchangeError::InvalidRequest)
+    ));
```

Comment:
Actually the error here (and in the other cases) is [returned by the client](https://github.com/eigerco/celestia-node-rs/blob/0c4b79ddd7db5ee72043eec003d126c66c56f369/node/src/exchange/client.rs#L68-L71). The request never goes to the server.

However we can keep the test.

Reply:
true, added mostly for completeness sake